Header checking

lindyhopchris opened the pull request:
Initial pass at resolving issue #68 

Interface changes:
- New interface `HeaderCheckerInterface`
- Modified interface `ParametersCheckerInterface` so that it only checks parameters.
- New interface `CombinedCheckerInterface` that checks both at once. (Not fussy about name.)

Implementation changes:
- `RestrictiveParameterChecker` now only checks parameters.
- `RestrictiveHeaderChecker` checks headers using same implementation as old `RestrictiveParameterChecker`.
- `RestrictiveChecker` checks both and keeps same constructor as old `RestrictiveParamterChecker`. I.e. people who want to continue using the previous way of doing checking use this new object instead.

Tests updated.

lindyhopchris committed:
Split parameter checker in two
Split the parameter checker into a HeaderCheckerInterface and ParameterCheckerInterface.
This allows framework implementations to do either checking at different at different places. For implementations that want to check both at once, the CombinedCheckerInterface can be used.

lindyhopchris committed:
Split RestrictiveParameterChecker in two.
Add RestrictiveHeaderChecker and reduced RestrictiveParameterChecker so that it only checks parameters.
Added RestrictiveChecker that is the equivalent of the previous RestrictiveParameterChecker - i.e. it checks both and has the same constructor arguments.

neomerx commented on src/Contracts/Parameters/CombinedCheckerInterface.php:
```diff
@@ -0,0 +1,19 @@
+ * Interface CombinedCheckerInterface
+ * @package Neomerx\JsonApi
+ */
+interface CombinedCheckerInterface extends HeaderCheckerInterface, ParameterCheckerInterface
```
Wouldn't it be more reasonable to think that ParametersChecker is (HeadersChecker and QueryChecker)? Yes, it might break compatibility but I was thinking moving to v0.6 from v0.5.13 anyways. Feel free to rename with those or better names.

lindyhopchris replied:
Yep I like `ParametersCheckerInterface` = `HeadersCheckerInterface` + `QueryCheckerInterface`. Will make that change.

neomerx commented on src/Parameters/RestrictiveChecker.php:
```diff
@@ -0,0 +1,89 @@
+        array $pagingParameters = null,
+        array $filteringParameters = null
+    ) {
+        $this->headerChecker = new RestrictiveHeaderChecker(
```
Please don't create them internally with `new` it makes testing more difficult and makes it less modular. You can pass `Interfaces` instead and add method to `Factory` that creates it with default implementations.

lindyhopchris replied:
Will do

neomerx commented:
And one more, please post to `develop` :wink:

lindyhopchris commented:
Will make changes and resubmit this to the develop branch.

lindyhopchris committed:
Update interface names.
Now `ParametersCheckerInterface` = `HeadersCheckerInterface` + `QueryCheckerInterface`.
Also updated the implementations of these interfaces so that they use the same names, i.e.
`RestrictiveParametersChecker` = `RestrictiveHeadersChecker` + `RestrictiveQueryChecker`

lindyhopchris committed:
Factory updates.
